pkg/agent/storage/storage.go

```diff
@@ -0,0 +1,287 @@
+package storage
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/spiffe/spire/pkg/common/diskutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+)
+
+var (
+	ErrNotCached = errors.New("not cached")
+)
+
+type Storage interface {
+	// LoadSVID loads the SVID from storage. Returns ErrNotCached if the SVID
+	// does not exist in the cache.
+	LoadSVID() ([]*x509.Certificate, error)
+
+	// StoreSVID stores the SVID.
+	StoreSVID(certs []*x509.Certificate) error
+
+	// DeleteSVID deletes the SVID.
+	DeleteSVID() error
+
+	// LoadBundle loads the SVID from storage. Returns ErrNotCached if the
+	// bundle does not exist in the cache.
+	LoadBundle() ([]*x509.Certificate, error)
+
+	// StoreBundle stores the bundle.
+	StoreBundle(certs []*x509.Certificate) error
+}
+
+func Open(dir string) (Storage, error) {
+	legacySVID, legacySVIDTime, err := loadLegacySVID(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	legacyBundle, legacyBundleTime, err := loadLegacyBundle(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	data, dataTime, err := loadData(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	storeNow := false
+
+	if !legacySVIDTime.IsZero() && (dataTime.IsZero() || dataTime.Before(legacySVIDTime)) {
+		storeNow = true
+		data.SVID = legacySVID
+	}
+
+	if !legacyBundleTime.IsZero() && (dataTime.IsZero() || dataTime.Before(legacyBundleTime)) {
+		storeNow = true
+		data.Bundle = legacyBundle
+	}
+
+	if storeNow {
+		if err := storeData(dir, data); err != nil {
```

Comment:
So we create the new json file here. Are the legacy files deleted?

Reply:
No, we have to preserve (and continue to update) the legacy files for the 1.4.x cycle, to allow for agents that are downgraded back to 1.3.x to continue to operate.